File: src/dyn_re_test.cpp
#include <ros/ros.h>
#include <dynamic_reconfigure/server.h>

#include "dyn_re_test/DynReTestConfig.h"

dynamic_reconfigure::Server<dyn_re_test::DynReTestConfig> *dyn_re = NULL;

void DynReCB( dyn_re_test::DynReTestConfig &cfg, uint32_t lvl)
{
	if( lvl == ~0 )
	{
		std::cout << "---------------------------------" << std::endl
			<< "groups configured" << std::endl;
		//cfg.groups.test_group_1.state = false;
		//dyn_re->updateConfig(cfg);
	}

	std::cout << "---------------------------------" << std::endl
		<< "int_param:\t\t" << cfg.int_param << std::endl
		<< "double_param:\t\t" << cfg.double_param << std::endl
		<< "inf_double_param:\t" << cfg.inf_double_param << std::endl
		<< "str_param:\t\t" << cfg.str_param << std::endl
		<< "bool_param:\t\t" << ( cfg.bool_param ? "True" : "False" ) << std::endl
		<< "enum_param:\t\t" << cfg.enum_param << std::endl
		<< "-test_group_1:\t\t" << cfg.groups.test_group_1.state << std::endl
		<< "--group_1_int_param:\t" << cfg.groups.test_group_1.group_1_int_param << std::endl
		<< "-test_group_2:\t\t" << cfg.groups.test_group_2.state << std::endl
		<< "--group_2_int_param:\t" << cfg.groups.test_group_2.group_2_int_param << std::endl
		<< "-test_group_3:\t\t" << cfg.groups.test_group_3.state << std::endl
		<< "--group_3_int_param:\t" << cfg.groups.test_group_3.group_3_int_param << std::endl
		<< "-test_group_4:\t\t" << cfg.groups.test_group_4.state << std::endl
		<< "--group_4_int_param:\t" << cfg.groups.test_group_4.group_4_int_param << std::endl
		<< "-test_group_5:\t\t" << cfg.groups.test_group_5.state << std::endl
		<< "--test_group_5_1:\t\t" << cfg.groups.test_group_5.test_group_5_1.state << std::endl
		<< "---group_5_1_str_param:\t\t" << cfg.groups.test_group_5.test_group_5_1.group_5_1_str_param << std::endl
		<< "---group_5_1_bool_param:\t" << ( cfg.groups.test_group_5.test_group_5_1.group_5_1_bool_param ? "True" : "False" ) << std::endl
		<< "--test_group_5_2:\t\t" << cfg.groups.test_group_5.test_group_5_2.state << std::endl
		<< "---group_5_2_int_param:\t\t" << cfg.groups.test_group_5.test_group_5_2.group_5_2_int_param << std::endl
		<< "---group_5_2_double_param:\t" << cfg.groups.test_group_5.test_group_5_2.group_5_2_double_param << std::endl;
}

int main( int argc, char *argv[] )
{
	ros::init( argc, argv, "dyn_re_test" );

	dyn_re = new dynamic_reconfigure::Server<dyn_re_test::DynReTestConfig>;
	dynamic_reconfigure::Server<dyn_re_test::DynReTestConfig>::CallbackType dyn_re_cb = boost::bind( &DynReCB, _1, _2 );

	std::cout << "Running..." << std::endl;

	dyn_re->setCallback( dyn_re_cb );

	ros::spin( );

	return 0;
}

File: cfg/DynReTest.py
#!/usr/bin/env python

PACKAGE = "dyn_re_test"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

gen.add("int_param",        int_t,    0, "An Integer parameter", 50,  0, 100)
gen.add("double_param",     double_t, 0, "A double parameter (bounded)",    .5, 0,   1)
gen.add("inf_double_param", double_t, 0, "A double parameter (bounded)",    0.0)
gen.add("str_param",        str_t,    0, "A string parameter",  "Hello World")
gen.add("bool_param",       bool_t,   0, "A Boolean parameter",  True)

size_enum = gen.enum([ gen.const("Small",      int_t, 0, "A small constant"),
                       gen.const("Medium",     int_t, 1, "A medium constant"),
                       gen.const("Large",      int_t, 2, "A large constant"),
                       gen.const("ExtraLarge", int_t, 3, "An extra large constant")],
                     "An enum to set size")

gen.add("enum_param", int_t, 0, "A size parameter which is edited via an enum", 1, 0, 3, edit_method=size_enum)

group1 = gen.add_group("test_group_1", "tab")
group1.add("group_1_int_param", int_t, 0, "An Integer parameter in a group", 0, -1000, 1000)

group2 = gen.add_group("test_group_2", "hide")
group2.add("group_2_int_param", int_t, 0, "An Integer parameter in a group", 0, -1000, 1000)

group3 = gen.add_group("test_group_3", "collapse")
group3.add("group_3_int_param", int_t, 0, "An Integer parameter in a group", 0, -1000, 1000)

group4 = gen.add_group("test_group_4", "apply")
group4.add("group_4_int_param", int_t, 0, "An Integer parameter in a group", 0, -1000, 1000)

group5 = gen.add_group("test_group_5", "tab")
group5_1 = group5.add_group("test_group_5_1", "collapse", False)
group5_1.add("group_5_1_str_param",        str_t,    0, "A string parameter",  "Hello World2")
group5_1.add("group_5_1_bool_param",       bool_t,   0, "A Boolean parameter",  True)
group5_2 = group5.add_group("test_group_5_2", "collapse")
group5_2.add("group_5_2_int_param",        int_t,    0, "An Integer parameter", 50,  0, 100)
group5_2.add("group_5_2_double_param",     double_t, 0, "A double parameter (bounded)",    .5, 0,   1)

exit(gen.generate(PACKAGE, "dyn_re_test", "DynReTest"))
